src/coreclr/vm/multicorejitplayer.cpp

```diff
@@ -1349,53 +1106,149 @@ HRESULT MulticoreJitProfilePlayer::PlayProfile()
+                    count++;
+
+                    pCurBuf += currcdLen;
+                    curSize -= currcdLen;
+
+                    if (curSize == 0)
+                    {
+                        break;
+                    }
 
-                if (((unsigned)(pRec->lenModuleName
-                    + pRec->lenAssemblyName
-                    ) > (rcdLen - sizeof(ModuleRecord))) ||
-                    (m_moduleCount >= m_headerModuleCount))
+                    unsigned curdata1 = * (const unsigned *) pCurBuf;
+                    isMethod = RecorderInfo::isMethod(curdata1);
+                }
+                while (isMethod && count < MAX_WALKBACK);
+
+                _ASSERTE(count > 0);
+                if (count > 1)
                 {
-                    hr = COR_E_BADIMAGEFORMAT;
+                    MulticoreJitTrace(("Jit backwards %d methods",  count));
                 }
-                else
+
+                int i = count - 1;
+                for (; (SUCCEEDED(hr)) && i >= 0; --i)
                 {
-                    hr = HandleModuleRecord(pRec);
+                    pCurBuf -= sizes[i];
+
+                    unsigned curdata1 = * (const unsigned *) pCurBuf;
+                    unsigned curmoduleIndex = curdata1 & (MAX_MODULES - 1);
+                    unsigned curflags = curdata1 & METHOD_FLAGS_MASK;
+
+                    unsigned curdata2 = * (((const unsigned *) pCurBuf) + 1);
+                    unsigned cursignatureLength = curdata2 & MAX_SIGNATURE_LENGTH;
+
+                    hr = HandleMethodInfoRecord(curmoduleIndex, (BYTE *) (pCurBuf + sizeof(unsigned) * 2), cursignatureLength);
+
+                    if (SUCCEEDED(hr) && ShouldAbort(false))
+                    {
+                        hr = E_ABORT;
+                    }
                 }
-            }
-            else if (rcdTyp == MULTICOREJIT_JITINF_RECORD_ID)
-            {
-                int mCount = (rcdLen - sizeof(unsigned)) / sizeof(unsigned);
 
-                hr = HandleMethodRecord((unsigned *) (pBuffer + sizeof(unsigned)), mCount);
-            }
-            else if (rcdTyp == MULTICOREJIT_GENERICINF_RECORD_ID)
-            {
-                unsigned info = *(unsigned *)(pBuffer + sizeof(unsigned));
-                unsigned moduleIndex = info >> 24;
-                unsigned signatureLength = info & SIGNATURELENGTH_MASK;
-                hr = HandleGenericMethodRecord(moduleIndex, (BYTE *) (pBuffer + sizeof(unsigned) * 2), signatureLength);
+                m_stats.m_nWalkBack += (short) count;
+                m_stats.m_nFilteredMethods += (short) (i + 1);
+
+                rcdLen = nSize - curSize;
             }
             else
             {
-                hr = COR_E_BADIMAGEFORMAT;
-            }
+                unsigned moduleIndex = data1 & (MAX_MODULES - 1);
+                unsigned flags = data1 & METHOD_FLAGS_MASK;
+
+                unsigned signatureLength = data2 & MAX_SIGNATURE_LENGTH;
```

Comment:
Same comment as above regarding `& MAX_SIGNATURE_LENGTH` or `& (MAX_SIGNATURE_LENGTH - 1)`, could be removed